src/codec/brotli/encoder.rs

```diff
@@ -1,38 +1,58 @@
 use crate::{codec::Encode, util::PartialBuffer};
-use std::{fmt, io::Result};
+use std::{
+    fmt,
+    io::{Error, ErrorKind, Result},
+};
 
-use brotli2::{
-    raw::{CoStatus, Compress, CompressOp},
-    CompressParams,
+use brotli::enc::{
+    backward_references::BrotliEncoderParams,
+    encode::{
+        BrotliEncoderCompressStream, BrotliEncoderCreateInstance, BrotliEncoderOperation,
+        BrotliEncoderStateStruct,
+    },
+    StandardAlloc,
 };
 
 pub struct BrotliEncoder {
-    compress: Compress,
+    state: BrotliEncoderStateStruct<StandardAlloc>,
 }
 
 impl BrotliEncoder {
-    pub(crate) fn new(params: &CompressParams) -> Self {
-        let mut compress = Compress::new();
-        compress.set_params(params);
-        Self { compress }
+    pub(crate) fn new(params: BrotliEncoderParams) -> Self {
+        let mut state = BrotliEncoderCreateInstance(StandardAlloc::default());
+        state.params = params;
+        Self { state }
     }
 
     fn encode(
         &mut self,
         input: &mut PartialBuffer<&[u8]>,
         output: &mut PartialBuffer<&mut [u8]>,
-        op: CompressOp,
-    ) -> Result<CoStatus> {
-        let mut in_buf = input.unwritten();
+        op: BrotliEncoderOperation,
+    ) -> Result<bool> {
+        let in_buf = input.unwritten();
         let mut out_buf = output.unwritten_mut();
 
-        let original_input_len = in_buf.len();
-        let original_output_len = out_buf.len();
-
-        let status = self.compress.compress(op, &mut in_buf, &mut out_buf)?;
+        let mut input_len = 0;
+        let mut output_len = 0;
 
-        let input_len = original_input_len - in_buf.len();
-        let output_len = original_output_len - out_buf.len();
+        let status = if BrotliEncoderCompressStream(
+            &mut self.state,
+            op,
+            &mut in_buf.len(),
+            in_buf,
+            &mut input_len,
+            &mut out_buf.len(),
+            &mut out_buf,
+            &mut output_len,
+            &mut None,
+            &mut |_, _, _, _| (),
+        ) <= 0
+        {
+            return Err(Error::new(ErrorKind::Other, "brotli error"));
+        } else {
+            self.state.available_out_ == 0
```

Comment:
This is supposed to be `BrotliEncoderIsFinished`, but it doesn't work if there are multiple streams and never enters `BROTLI_STREAM_FINISHED` state.
This may be a bug in `brotli` or fixable in `async-compression`, I'm not sure.

Reply:
This is because the `FINISHED` state is not hit when calling `flush`, changing this to

```rust
if op == BrotliEncoderOperation::BROTLI_OPERATION_FLUSH {
    BrotliEncoderHasMoreOutput(&self.state) == 0
} else {
    BrotliEncoderIsFinished(&self.state) != 0
}
```

works, though maybe it'd be better to return both those checks as flags that can be checked in `flush` and `finish`.

Reply:
I just moved the checks down to `flush` and `finish`.